tools/windows/DatadogAgentInstaller/CustomActions/PythonDistributionCustomAction.cs

```diff
@@ -225,23 +255,70 @@ private static ActionResult RunPythonScript(ISession session, string script)
             return ActionResult.Success;
         }
 
-        public static ActionResult RunPostInstPythonScript(Session session)
+        private ActionResult RunPostInstPythonScript()
         {
-            ISession sessionWrapper = new SessionWrapper(session);
             // check if INSTALL_PYTHON_THIRD_PARTY_DEPS property is set
-            var installPythonThirdPartyDeps = sessionWrapper.Property("INSTALL_PYTHON_THIRD_PARTY_DEPS");
+            var installPythonThirdPartyDeps = _session.Property("INSTALL_PYTHON_THIRD_PARTY_DEPS");
             if (string.IsNullOrEmpty(installPythonThirdPartyDeps) || installPythonThirdPartyDeps != "1")
             {
-                sessionWrapper.Log("Skipping installation of third-party Python deps. Set INSTALL_PYTHON_THIRD_PARTY_DEPS=1 to enable this feature.");
+                _session.Log("Skipping installation of third-party Python deps. Set INSTALL_PYTHON_THIRD_PARTY_DEPS=1 to enable this feature.");
                 return ActionResult.Success;
             }
+            return RunPythonScript(_session, "post.py");
 
-            return RunPythonScript(sessionWrapper, "post.py");
+        }
+
+        private ActionResult RunPreRemovePythonScript()
+        {
+            // add the .post_python_installed_packages.txt to the rollback data store
+            // This means the file can be restored on failure and still used to create a diff file for next retry
+            // Helps prevents silent failure of third party integration feature. Where there is no post file so it silenly continues.
+            var pythonPackagesFile = Path.Combine(_session.Property("APPLICATIONDATADIRECTORY"), "protected", ".post_python_installed_packages.txt");
+            // check that file path exists
+            if (!System.IO.File.Exists(pythonPackagesFile))
+            {
+                _session.Log($"File {pythonPackagesFile} does not exist, skipping rollback data store addition.");
+            }
+            else
+            {
+                _rollbackDataStore.Add(new FileStorageRollbackData(pythonPackagesFile));
+            }
+
+            var result = RunPythonScript(_session, "pre.py");
+            _rollbackDataStore.Store();
```

Comment:
use finally block to ensure rollback data is stored in case of exception/error

```suggestion
            try
            {
                return RunPythonScript(_session, "pre.py");
            }
            finally
            {
                _rollbackDataStore.Store();
            }
            
```